### PR TITLE
Surveys: Add multi-select as an input type

### PR DESCRIPTION
**Pull request by NWM-OS**
#ENA-510
#3192

### CODE REVIEW

File: waltz-data/src/main/java/com/khartec/waltz/data/survey/SurveyQuestionResponseDao.java [hunk 1]
```diff
@@ -122,29 +150,80 @@ public void saveResponse(SurveyInstanceQuestionResponse response) {
...
-
-        records.stream()
-                .forEach(r -> r.setSurveyInstanceId(targetSurveyInstanceId));
-
-        return dsl.batchInsert(records)
-                .execute();
+                .fetchInto(SURVEY_QUESTION_RESPONSE)
+                .stream()
+                .peek(r -> r.setSurveyInstanceId(targetSurveyInstanceId))
+                .peek(r -> r.changed(true))
```
Comment (davidwatkins73) [inline waltz-data/src/main/java/com/khartec/waltz/data/survey/SurveyQuestionResponseDao.java]: I thought using `peek` was mostly to support debug/visibility and using it to modify elements in a stream was generally frowned upon.  

Is this not the same as `.map(r -> { r.setSurveyInstance(t); r.changed(true); return r; })`  ?

File: waltz-data/src/main/java/com/khartec/waltz/data/survey/SurveyQuestionResponseDao.java [hunk 2]
```diff
@@ -122,29 +150,80 @@ public void saveResponse(SurveyInstanceQuestionResponse response) {
...
+                .collect(toList());
+
+        List<SurveyQuestionListResponseRecord> listResponseRecords = dsl
+                .selectFrom(SURVEY_QUESTION_LIST_RESPONSE)
+                .where(SURVEY_QUESTION_LIST_RESPONSE.SURVEY_INSTANCE_ID.eq(sourceSurveyInstanceId))
+                .fetchInto(SURVEY_QUESTION_LIST_RESPONSE)
+                .stream()
+                .peek(r -> r.setSurveyInstanceId(targetSurveyInstanceId))
+                .peek(r -> r.changed(true))
```
Comment (davidwatkins73) [inline waltz-data/src/main/java/com/khartec/waltz/data/survey/SurveyQuestionResponseDao.java]: ditto re:`peek`

File: waltz-data/src/main/java/com/khartec/waltz/data/survey/SurveyQuestionResponseDao.java [hunk 3]
```diff
@@ -122,29 +150,80 @@ public void saveResponse(SurveyInstanceQuestionResponse response) {
...
+
+                if (! lrs.isEmpty()) {
+                    List<SurveyQuestionListResponseRecord> listResponses =
+                            IntStream.range(0, lrs.size())
+                            .mapToObj(i -> {
+                                SurveyQuestionListResponseRecord rec = new SurveyQuestionListResponseRecord();
+                                rec.setSurveyInstanceId(response.surveyInstanceId());
+                                rec.setQuestionId(response.questionResponse().questionId());
+                                rec.setResponse(lrs.get(i));
+                                rec.setPosition(i + 1);
```
Comment (davidwatkins73) [inline waltz-data/src/main/java/com/khartec/waltz/data/survey/SurveyQuestionResponseDao.java]: See other comment about possibly using an `AtomicInt` to remove need for the `IntStream`

**Comment by davidwatkins73** [comment]
Hi, 
added some review comments, but could you elaborate a bit on the need for the response_list_concat field....  Looks like something it would be easy for us to forget to update.

**Comment by rovats** [comment]
Hi Dave,

I documented the approach in the issue:
https://github.com/khartec/waltz/issues/3192

list_response_concat is only to simplify export queries, rather than having to concat multiple records on the fly (which is DB specific, eg: for xml path), it’s computed on input.

File: waltz-data/src/main/java/com/khartec/waltz/data/survey/SurveyQuestionResponseDao.java (hunk 1 above)
Comment (rovats) [reply waltz-data/src/main/java/com/khartec/waltz/data/survey/SurveyQuestionResponseDao.java]: I used map initially, but IntelliJ suggested peek, but I think I’ll change back to use .map or .forEach.

File: waltz-data/src/main/ddl/liquibase/db.changelog-1.23.xml [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+            <column name="survey_instance_id"
+                    type="${id.type}">
+                <constraints nullable="false"/>
+            </column>
+            <column name="question_id"
+                    type="${id.type}">
+                <constraints nullable="false"/>
+            </column>
+            <column name="response"
+                    type="nvarchar(max)">
```
[the remark this answers is not in the document]
Comment (rovats) [reply waltz-data/src/main/ddl/liquibase/db.changelog-1.23.xml]: Will create a new property `maxvarchar.type` for this:

```
<property name="maxvarchar.type" value="nvarchar(max)" dbms="mssql"></property>
<property name="maxvarchar.type" value="LONGVARCHAR" dbms="mysql"></property>
<property name="maxvarchar.type" value="varchar" dbms="postgresql"></property>
<property name="maxvarchar.type" value="LONGVARCHAR" dbms="h2"></property>
```

**Comment by davidwatkins73** [comment]
thanks for making the updates!

**Pull request merged**